### ISSUE TITLE
Non-unique serialisation ids

### ISSUE THREAD

**Issue by Bathtor** [cda-group/arcon#94]
```rust
impl<A: ArconType> UnsafeSerde<A> {
    const SID: kompact::prelude::SerId = 26;
}
```
This is a really bad idea. Assigning the same serialisation id to multiple concrete serialisers (i.e. one for every generic type) can end you up with someone trying to deserialise data of type B with a serialiser for  type A and getting *some* weird as shit value by accident without error.

This may work ok for protobuf stuff assuming it maintains its own internal ids (which I'm not sure it does, tbh), but for abomonation this is definitely a no-no.

It would be better to generate serialisation id for every type `A` and have `ReliableSerde<A>` use `A::SER_ID` instead, so that it can't be confused with `B`.

Kompact has the following trait for this purpose to be implemented on `A` *not* `ReliableSerde<A>`:
```rust
pub trait SerialisationId {
    const SER_ID: SerId;
}
```
Alternatively you can make add a `SER_ID` field to `ArconType`. 

We will need a way to generate those ids reliably and consistently from the Arc compiler toolchain at some point.

**Comment by Max-Meldrum** [comment]
yeah, good point. I guess the current approach works under the assumption that the system is always set up correctly ^^

I'm wondering if we might need a version identifier in addition to a SER_ID.  First of all, we need to know whether if it is Protobuf or Abomonation. Secondly, it would be good to check that their versions match.

If we don't have predefined identifiers for Protobuf/Abomonation, how  would the following logic work in the Arcon node? 

```rust
fn receive_network(&mut self, msg: NetMessage) {
        let arcon_msg: ArconResult<ArconMessage<IN>> = match *msg.ser_id() {
            ReliableSerde::<IN>::SER_ID => msg
                .try_deserialise::<ArconMessage<IN>, ReliableSerde<IN>>()
                .map_err(|_| arcon_err_kind!("Failed to unpack reliable ArconMessage")),
            UnsafeSerde::<IN>::SER_ID => msg
                .try_deserialise::<ArconMessage<IN>, UnsafeSerde<IN>>()
                .map_err(|_| arcon_err_kind!("Failed to unpack unreliable ArconMessage")),
            _ => panic!("Unexpected deserialiser"),
        };
}
```

Would ``IN: ArconType`` have to impl ``UNSAFE_SER_ID``,  ``RELIABLE_SER_ID``  and match on ``msg.ser_id()``

**Comment by Bathtor** [comment]
>I'm wondering if we might need a version identifier in addition to a SER_ID. First of all, we need to know whether if it is Protobuf or Abomonation. Secondly, it would be good to check that their versions match.

Well, either we generate a new serialisation id for each version (basically folding the version into the serialisation id) or we prepend the version to the serialised data. I think the latter approach is going to be better long term, because it will allow us to generate custom conversion code for when we are reading old version from snapshots into code running a newer version. This is also basically the protobuf approach, so I think it already does this internal, if I'm not mistaken.

> If we don't have predefined identifiers for Protobuf/Abomonation, how would the following logic work in the Arcon node?

You can simply forward the `SER_ID` of the serialised type `T` in `impl Deserialiser<T> for ReliableSerde<T>`.

See https://github.com/kompics/kompact/blob/master/core/src/serialisation/serde_serialisers.rs as an example of this approach.

**Comment by Max-Meldrum** [comment]
Looking at this now

**Comment by Max-Meldrum** [comment]
Resolved through #101
